src/NServiceBus.Persistence.CosmosDB/Saga/SagaPersister.cs

```diff
@@ -44,30 +47,64 @@ public Task Update(IContainSagaData sagaData, SynchronizedStorageSession session
 
             var responseMessage = await container.ReadItemStreamAsync(sagaId.ToString(), partitionKey).ConfigureAwait(false);
 
-            if (responseMessage.StatusCode == HttpStatusCode.NotFound || responseMessage.Content == null)
+            var sagaStream = responseMessage.Content;
+
+            var sagaNotFound = responseMessage.StatusCode == HttpStatusCode.NotFound || sagaStream == null;
+
+            if (sagaNotFound && !migrationModeEnabled)
             {
                 return default;
             }
 
-            using (var streamReader = new StreamReader(responseMessage.Content))
+            if (sagaNotFound && migrationModeEnabled && !isGeneratedSagaId)
+            {
+                var query = $@"SELECT TOP 1 * FROM c WHERE c[""{MetadataExtensions.MetadataKey}""][""{MetadataExtensions.SagaDataContainerMigratedSagaIdMetadataKey}""] = '{sagaId}'";
+                var queryDefinition = new QueryDefinition(query);
+                var queryStreamIterator = container.GetItemQueryStreamIterator(queryDefinition);
+
+                var iteratorResponse = await queryStreamIterator.ReadNextAsync().ConfigureAwait(false);
+
+                iteratorResponse.EnsureSuccessStatusCode();
+
+                using (var streamReader = new StreamReader(iteratorResponse.Content))
+                {
+                    using (var jsonReader = new JsonTextReader(streamReader))
+                    {
+                        var iteratorResult = JObject.Load(jsonReader);
+
+                        var documents = iteratorResult["Documents"] as JArray;
+
+                        if (!documents.HasValues)
+                        {
+                            return default;
+                        }
+
+                        sagaStream = new MemoryStream(Encoding.UTF8.GetBytes(documents[0].ToString()));
```

Comment:
We can skim stream allocation and converting to bytes by converting JObject to TSagaData and returning early.
```suggestion
                        var sagaData = documents[0].ToObject<TSagaData>();
                        context.Set($"cosmos_etag:{sagaData.Id}", responseMessage.Headers.ETag);
                        return sagaData;
```